File: 0036-valid-sudoku/0036-valid-sudoku.py
from typing import List

class Solution:
    # Constructor
    def __init__(self):
        self.n = 9
        self.rows = [set() for _ in range(self.n)]
        self.columns = [set() for _ in range(self.n)]
        self.boxes = [set() for _ in range(self.n)]
    
    # Valid sudoku
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for r in range(self.n):
            for c in range(self.n):
                # Value of the board
                value = board[r][c]
                if value == ".":
                    continue
                
                # Values in rows
                if value in self.rows[r]:
                    return False
                self.rows[r].add(value)

                # Values in columns
                if value in self.columns[c]:
                    return False
                self.columns[c].add(value)

                # Values in boxes
                boxIndex = (r // 3) * 3 + (c // 3)
                if value in self.boxes[boxIndex]:
                    return False
                self.boxes[boxIndex].add(value)
        
        return True

File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public ArrayList<Set<Character>> rows = new ArrayList<>();
    public ArrayList<Set<Character>> cols = new ArrayList<>();
    public ArrayList<Set<Character>> boxes = new ArrayList<>();


    
    public Solution() {
        // Initialize the ArrayLists with 9 empty sets
        for (int i = 0; i < 9; i++) {
            rows.add(new HashSet<>());
            cols.add(new HashSet<>());
            boxes.add(new HashSet<>());
        }
    }
    public boolean isValidSudoku(char[][] board) {
        //1. For loop 
        for(int i = 0;i<9;i++){
            for(int j = 0;j<9;j++){
                if(board[i][j] == '.'){
                    continue;
                }
                
                //check if its in rows 
                if(rows.get(i).contains(board[i][j])){
                    return false;
                }
                rows.get(i).add(board[i][j]);
                
                //check if its in cols 
                if(cols.get(j).contains(board[i][j])){
                    return false;
                }
                cols.get(j).add(board[i][j]);
                
                //check if its in boxes
                if(boxes.get((int)(i/3)*3 + (int)(j/3)).contains(board[i][j])){
                    return false;
                }
                boxes.get((int)(i/3)*3 + (int)(j/3)).add(board[i][j]);
                
                
                
                
            }
        }
        return true;
    }
}